go/interpreter/lfvm/memory.go

```diff
@@ -68,47 +68,48 @@ func (m *Memory) ExpansionCosts(size uint64) tosca.Gas {
 	return fee
 }
 
-func (m *Memory) EnsureCapacity(offset, size uint64, c *context) error {
-	if size <= 0 {
+// expandMemory tries to expand memory to the given size.
+// If the memory is already large enough or size is 0, it does nothing.
+// If not enough gas in the context or an overflow occurs when adding offset and
+// size, it returns an error.
+func (m *Memory) expandMemory(offset, size uint64, c *context) error {
+	if size == 0 {
 		return nil
 	}
+
 	needed := offset + size
 	// check overflow
 	if needed < offset {
 		c.signalError()
 		return errGasUintOverflow
 	}
-	if m.Len() < needed {
-		needed = toValidMemorySize(needed)
-		fee := m.ExpansionCosts(needed)
-		if !c.useGas(fee) {
-			c.status = statusOutOfGas
-			return errOutOfGas
-		}
-		m.total_memory_cost += fee
-		m.store = append(m.store, make([]byte, needed-m.Len())...)
+
+	fee := m.ExpansionCosts(needed)
+	if !c.useGas(fee) {
+		c.status = statusOutOfGas
+		return errOutOfGas
 	}
+
+	m.expandMemoryWithoutCharging(needed, fee)
 	return nil
 }
 
-func (m *Memory) EnsureCapacityWithoutGas(size uint64) {
-	if size <= 0 {
-		return
-	}
-	if m.Len() < size {
-		size = toValidMemorySize(size)
-		fee := m.ExpansionCosts(size)
-		m.total_memory_cost += fee
-		m.store = append(m.store, make([]byte, size-m.Len())...)
+// expandMemoryWithoutCharging expands the memory to the given size without checking gas.
+func (m *Memory) expandMemoryWithoutCharging(needed uint64, fee tosca.Gas) {
```

Comment:
Accepting the expansion `fee` from an outside caller is a violation of the encapsulation principle. I would drop this parameter and re-compute the fee inside the function. The performance impact of this should be negligible.